tests/wpt/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+// in clippy.
+#[allow(unused)]
+fn execute_test_file(path: &Path) {
+    let dir = path.parent().unwrap();
+    let wpt_path = PathBuf::from(
+        std::env::var("WPT_ROOT").expect("Could not find WPT_ROOT environment variable"),
+    );
+    let (mut context, logger) = create_context(&wpt_path);
+    let test_done = TestCompletion::new();
+
+    // Insert the logger to be able to access the logs after the test is done.
+    context.insert_data(logger.clone());
+    context.insert_data(test_done.clone());
+
+    let function = result_callback__
+        .into_js_function_copied(&mut context)
+        .to_js_function(context.realm());
+    context
+        .register_global_property(js_str!("result_callback__"), function, Attribute::all())
+        .expect("Could not register result_callback__");
+    context
+        .eval(Source::from_bytes(
+            b"add_result_callback(result_callback__);",
+        ))
+        .expect("Could not eval add_result_callback");
+
+    let function = complete_callback__
+        .into_js_function_copied(&mut context)
+        .to_js_function(context.realm());
+    context
+        .register_global_property(js_str!("complete_callback__"), function, Attribute::all())
+        .expect("Could not register complete_callback__");
+    context
+        .eval(Source::from_bytes(
+            b"add_completion_callback(complete_callback__);",
+        ))
+        .expect("Could not eval add_completion_callback");
+
+    // Load the test.
+    let source = TestSuiteSource::new(path);
+    for script in source.scripts().expect("Could not get scripts") {
+        // Resolve the source path relative to the script path, but under the wpt_path.
+        let script_path = Path::new(&script);
+        let path = if script_path.is_relative() {
+            dir.join(script_path)
+        } else {
+            wpt_path.join(script_path.to_string_lossy().trim_start_matches('/'))
+        };
+
+        if path.exists() {
+            let source = Source::from_filepath(&path).expect("Could not parse source.");
+            if let Err(err) = context.eval(source) {
+                panic!("Could not eval script, path = {path:?}, err = {err:?}");
+            }
+        } else {
+            panic!("Script does not exist, path = {path:?}");
+        }
+    }
+    context
+        .eval(source.source())
+        .expect("Could not evaluate source");
+    context.run_jobs();
+
+    // Done()
+    context
+        .eval(Source::from_bytes(b"done()"))
+        .expect("Done unexpectedly threw an error.");
+
+    let start = std::time::Instant::now();
+    while !test_done.is_done() {
+        context.run_jobs();
+
+        assert!(
+            start.elapsed().as_secs() < 10,
+            "Test did not complete in 10 seconds."
+        );
+    }
+}
+
+/// Test the console with the WPT test suite.
+#[cfg(not(clippy))]
+#[rstest::rstest]
+fn console(
+    #[base_dir = "${WPT_ROOT}"]
```

Comment:
I'm thinking that it would be better to add an internal tool to clone the wpt repo on a certain commit hash in the project root, then change `WPT_ROOT` to point to that instead. Makes it easier to keep track of the latest tested hash.

Reply:
I think `build.rs` works when you absolutely need to download something for your build, but in our case we don't want to always download the repo, and putting that in the `build.rs` will force everyone to download that, even if they're not working on `boa_runtime`.

Reply:
Another disadvantage I see is that if you wanted to bump the commit hash of wpt, you would be forced to recompile the crate, even if you didn't change anything about the code.

Reply:
Separating the crate does seem like a nice solution for the optional tests. This doesn't really solve the recompilation problem about `build.rs` though

Reply:
> because build.rs only tracks if tests_wpt have changed

It checks that the SHA at the HEAD of the tests_wpt/ folder is different. If it is, it will fetch and checkout the new rev.

Manually doing it on my computer (changing to a previous SHA then running build then checking the SHA):

<img width="1840" alt="CleanShot 2025-07-01 at 17 01 40@2x" src="https://github.com/user-attachments/assets/204a5039-f4c3-4c04-87ef-8d3ecbd47ab3" />

Then running `cargo build` in the root which is instant, then in the project which takes about 13 seconds, then in the project again:

<img width="1840" alt="image" src="https://github.com/user-attachments/assets/c618ecb0-2014-4b14-8255-61732c642149" />

Reply:
That's not exactly what I mean. If you change the hash in `tests_wpt_config.toml` without touching anything inside `tests_wpt`, does the `build.rs` detect that it needs to fetch the new hash?

Reply:
Seems the `build` script is always run (even if nothing changes), but doesn't do anything if there's no change in the hash or the repo.